moveit_setup_assistant/src/widgets/controllers_widget.cpp

```diff
@@ -0,0 +1,834 @@
+    groups.push_back(group_it->name_);
+  }
+
+  // Set the available groups (left box)
+  joint_groups_widget_->setAvailable(groups);
+
+  // Set the selected groups (right box)
+  joint_groups_widget_->setSelected(this_controller->joints_);
+
+  // Set the title
+  joint_groups_widget_->title_->setText(
+      QString("Edit '").append(QString::fromUtf8(this_controller->name_.c_str())).append("' Joints groups collection"));
+
+  // Remember what is currently being edited so we can later save changes
+  current_edit_controller_ = this_controller->name_;
+}
+
+// ******************************************************************************************
+// Delete a controller
+// ******************************************************************************************
+void ControllersWidget::deleteController()
+{
+  std::string controller_name = current_edit_controller_;
+
+  if (controller_name.empty())
+  {
+    QTreeWidgetItem* item = controllers_tree_->currentItem();
+    // Check that something was actually selected
+    if (item == NULL)
+      return;
+
+    // Get the user custom properties of the currently selected row
+    int type_ = item->data(0, Qt::UserRole).value<int>();
+    if (type_ == 0)
+      controller_name = item->text(0).toUtf8().constData();
+  }
+  else
+    current_edit_controller_.clear();
+  if (controller_name.empty())
+    return;
+
+  // Confirm user wants to delete controller
+  if (QMessageBox::question(this, "Confirm Controller Deletion",
+                            QString("Are you sure you want to delete the controller '").append(controller_name.c_str()),
+                            QMessageBox::Ok | QMessageBox::Cancel) == QMessageBox::Cancel)
+  {
+    return;
+  }
+  // delete actual controller
+  for (std::vector<Controller>::iterator controller_it = config_data_->controllers_config_.begin();
+       controller_it != config_data_->controllers_config_.end(); ++controller_it)
+  {
+    if (controller_it->name_ == controller_name)  // string match
+    {
+      config_data_->controllers_config_.erase(controller_it);  // convert to pointer from iterator
+      break;                                                   // we are done searching
+    }
+  }
+
+  // Switch to main screen
+  showMainScreen();
+
+  // Reload main screen table
+  loadControllersTree();
+}
+
+// ******************************************************************************************
+// Create a new, controller
+// ******************************************************************************************
+void ControllersWidget::addController()
+{
+  // New Controller
+  adding_new_controller_ = true;
+
+  // Load the data
+  loadControllerScreen(NULL);  // NULL indicates this is a new controller, not an existing one
+  changeScreen(2);
+}
+
+// ******************************************************************************************
+// Add a Follow Joint Trajectory action Controller for each Planning Group
```

Comment:
"Add a FollowJointTrajectory action controller from ros_control for each Planning Group"